collector/indices_mappings.go

```diff
@@ -0,0 +1,177 @@
+}
+
+// NewIndicesMappings defines Indices IndexMappings Prometheus metrics
+func NewIndicesMappings(logger log.Logger, client *http.Client, url *url.URL) *IndicesMappings {
+	subsystem := "indices_mappings_stats"
+
+	return &IndicesMappings{
+		logger: logger,
+		client: client,
+		url:    url,
+
+		up: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: prometheus.BuildFQName(namespace, subsystem, "up"),
+			Help: "Was the last scrape of the ElasticSearch Indices Mappings endpoint successful.",
+		}),
+		totalScrapes: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, subsystem, "total_scrapes"),
+			Help: "Current total ElasticSearch Indices Mappings scrapes.",
+		}),
+		jsonParseFailures: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, subsystem, "json_parse_failures"),
+			Help: "Number of errors while parsing JSON.",
+		}),
+		metrics: []*indicesMappingsMetric{
+			{
+				Type: prometheus.GaugeValue,
+				Desc: prometheus.NewDesc(
+					prometheus.BuildFQName(namespace, subsystem, "field_count"),
+					"Current number fields within cluster.",
+					defaultIndicesMappingsLabels, nil,
+				),
+				Value: func(indexMapping IndexMapping) float64 {
+					return countFieldsRecursive(indexMapping.Mappings.Properties, 0)
+				},
+			},
+		},
+	}
+}
+
+func countFieldsRecursive(properties IndexMappingProperties, fieldCounter float64) float64 {
+	// iterate over all properties
+	for _, property := range properties {
+		if property.Type != nil {
+			// property has a type set - counts as a field
+			fieldCounter++
+
+			// iterate over all fields of that property
+			for _, field := range property.Fields {
+				// field has a type set - counts as a field
+				if field.Type != nil {
+					fieldCounter++
+				}
+			}
+		}
+
+		// count recursively in case the property has more properties
+		if property.Properties != nil {
+			fieldCounter = +countFieldsRecursive(property.Properties, fieldCounter)
+		}
+	}
+
+	return fieldCounter
+}
+
+// Describe add Snapshots metrics descriptions
+func (im *IndicesMappings) Describe(ch chan<- *prometheus.Desc) {
+	for _, metric := range im.metrics {
+		ch <- metric.Desc
+	}
+
+	ch <- im.up.Desc()
+	ch <- im.totalScrapes.Desc()
+	ch <- im.jsonParseFailures.Desc()
+}
+
+func (im *IndicesMappings) getAndParseURL(u *url.URL) (*IndicesMappingsResponse, error) {
+	res, err := im.client.Get(u.String())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get from %s://%s:%s%s: %s",
+			u.Scheme, u.Hostname(), u.Port(), u.Path, err)
+	}
+
+	defer func() {
+		err = res.Body.Close()
+		if err != nil {
+			_ = level.Warn(im.logger).Log(
+				"msg", "failed to close http.Client",
+				"err", err,
+			)
+		}
+	}()
+
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("HTTP Request failed with code %d", res.StatusCode)
+	}
+
+	var imr IndicesMappingsResponse
+	if err := json.NewDecoder(res.Body).Decode(&imr); err != nil {
```

Comment:
This is not a big deal, but it would be cleaner to use `json.Unmarshal()` instead. The decoder isn't guaranteed to read the whole body of the response and could leave the connection lingering open and unusable.  That also means you can explicitly close the body after reading the whole thing to ensure that the connections get returned to the pool for reuse. 

https://changelog.com/gotime/141#transcript-111
https://pkg.go.dev/net/http?tab=doc#Response see the notes on the Body property of the Response here.